btree/btree.c

```diff
@@ -1501,6 +1502,263 @@ static void generic_move(struct nd *src, struct nd *tgt,
+}
+
+/**
+ * Calculates the credit needed to update @nr nodes and adds this credit to
+ * @accum.
+ */
+static void btree_node_update_credit(struct m0_be_tx_credit *accum,
+				     m0_bcount_t             nr)
+{
+	struct m0_be_tx_credit cred = {};
+
+	/* @todo: why multiplication with 2 is needed? */
+	m0_be_tx_credit_mac(&cred, max_node_size, 2);
+
+	m0_be_tx_credit_mac(accum, &cred, nr);
+}
+
+/**
+ * Calculates the credit needed to create @nr nodes and adds this credit to
+ * @accum.
+ */
+static void m0_btree_create_node_credit(const struct m0_btree  *tree,
+					m0_bcount_t             nr,
+					struct m0_be_tx_credit *accum)
+{
+	struct m0_be_tx_credit cred = {};
+
+	btree_node_alloc_credit(tree, &cred);
+	/* @todo: is btree_node_update_credit needed? */
+	btree_node_update_credit(&cred, 1);
+	m0_be_tx_credit_mac(accum, &cred, nr);
+}
+
+/**
+ * Calculates the credit needed to move some records between two nodes and
+ * adds this credit to @accum.
+ */
+static void btree_move_key_value_credit(struct m0_be_tx_credit *accum,
+					enum node_threshold th)
+{
+	/*depending on node_threshold, calculate fraction which needs to be
+	moved to other node*/
+	m0_bcount_t size = max_node_size*th/100;
+
+	m0_be_tx_credit_mac(accum, &M0_BE_TX_CREDIT(1, size), 1);
+	btree_node_update_credit(accum, 1);
+}
+
+static void btree_insert_key_value_credit(const struct m0_btree  *tree,
+					  m0_bcount_t             ksize,
+					  m0_bcount_t             vsize,
+					  struct m0_be_tx_credit *accum)
+{
+	struct m0_be_tx_credit kv_update_cred;
+
+	kv_update_cred = M0_BE_TX_CREDIT(1, ksize + vsize);
+	btree_mem_alloc_credit(tree, ksize + vsize, accum);
+	m0_be_tx_credit_add(accum, &kv_update_cred);
+}
+
+static void btree_insert_leaf_node_credit(const struct m0_btree  *tree,
+					  m0_bcount_t             ksize,
+					  m0_bcount_t             vsize,
+					  struct m0_be_tx_credit *accum)
+{
+	m0_btree_create_node_credit(tree, 1, accum);
+	btree_move_key_value_credit(accum, MAX_MOVE);
+	btree_insert_key_value_credit(tree, ksize, vsize, accum);
+}
+
+static void btree_insert_internal_node_credit(const struct m0_btree  *tree,
+					      struct m0_be_tx_credit *accum)
+{
+	m0_btree_create_node_credit(tree, 1, accum);
+	btree_move_key_value_credit(accum, MAX_MOVE);
+	btree_insert_key_value_credit(tree, max_ksize, internal_vsize, accum);
+}
+
+static void btree_insert_new_root_credit(const struct m0_btree  *tree,
+					 struct m0_be_tx_credit *accum)
+{
+	struct m0_be_tx_credit cred = {};
+
+	m0_btree_create_node_credit(tree, 1, accum);
+	btree_move_key_value_credit(accum, ALL_MOVE);
+
+	/*credit for inserting 2 key-value at root node*/
+	btree_insert_key_value_credit(tree, max_ksize, internal_vsize, &cred);
+	m0_be_tx_credit_mac(accum, &cred, 2);	
+}
+
+/**
+ * Calculates total credits required to perform @nr insert operations
+ * @param nr     Number of insert operations.
+ * @param ksize  Key data size.
+ * @param vsize  Value data size.
+*/
```

Comment:
Please add comment:
@param accum the accumulated credit will be updated to have credit for insert operation.

Reply:
okay